### PR TITLE
fix #37359 Return a Diagonal matrix when Abstract type is a Number

### LINKED ISSUE #766: `\(::UniformScaling, ::Diagonal{T,Array{T,1}})` returns dense `Matrix` when `T` is abstract
As stated in the title, `\(::UniformScaling, ::Diagonal{T,Array{T,1}})` returns a dense `Matrix` when `T` is abstract. It seems that, ideally, a `Diagonal` matrix should be returned (as is the case when `T` is concrete). (This is for the case when `T <: Number` is `true`; I haven't looked at what happens when `T <: Number` is `false`.)
```julia
julia> versioninfo()
Julia Version 1.5.1
Commit 697e782ab8 (2020-08-25 20:08 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 8

julia> I \ Diagonal(ones(Integer, 4)) # T abstract --> returns Array
4×4 Array{Float64,2}:
 1.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0
 0.0  0.0  1.0  0.0
 0.0  0.0  0.0  1.0

julia> I \ Diagonal(ones(Int, 4)) # T concrete --> returns Diagonal
4×4 Diagonal{Float64,Array{Float64,1}}:
 1.0   ⋅    ⋅    ⋅ 
  ⋅   1.0   ⋅    ⋅ 
  ⋅    ⋅   1.0   ⋅ 
  ⋅    ⋅    ⋅   1.0
```

### LINKED ISSUE #37359: `\(::UniformScaling, ::Diagonal{T,Array{T,1}})` returns dense `Matrix` when `T` is abstract
As stated in the title, `\(::UniformScaling, ::Diagonal{T,Array{T,1}})` returns a dense `Matrix` when `T` is abstract. It seems that, ideally, a `Diagonal` matrix should be returned (as is the case when `T` is concrete). (This is for the case when `T <: Number` is `true`; I haven't looked at what happens when `T <: Number` is `false`.)
```julia
julia> versioninfo()
Julia Version 1.5.1
Commit 697e782ab8 (2020-08-25 20:08 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 8

julia> I \ Diagonal(ones(Integer, 4)) # T abstract --> returns Array
4×4 Array{Float64,2}:
 1.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0
 0.0  0.0  1.0  0.0
 0.0  0.0  0.0  1.0

julia> I \ Diagonal(ones(Int, 4)) # T concrete --> returns Diagonal
4×4 Diagonal{Float64,Array{Float64,1}}:
 1.0   ⋅    ⋅    ⋅ 
  ⋅   1.0   ⋅    ⋅ 
  ⋅    ⋅   1.0   ⋅ 
  ⋅    ⋅    ⋅   1.0
```

### PR DESCRIPTION
**Pull request by tantei3**
Closes JuliaLang/LinearAlgebra.jl#766

### CODE REVIEW

File: stdlib/LinearAlgebra/test/diagonal.jl [hunk 1]
```diff
@@ -726,4 +726,7 @@ end
     @test dot(zeros(Int32, 0), Diagonal(zeros(Int, 0)), zeros(Int16, 0)) === 0
 end
 
+# issue 37359
+@test isa(I \ Diagonal(ones(Integer, n)), Diagonal)
```
Comment (fredrikekre) [inline stdlib/LinearAlgebra/test/diagonal.jl:730]: Would be good if the test, in addition, tested the value of the computation.

Comment (tantei3) [reply stdlib/LinearAlgebra/test/diagonal.jl]: Sure, have updated the tests

File: stdlib/LinearAlgebra/test/diagonal.jl [hunk 2]
```diff
@@ -726,4 +726,12 @@ end
     @test dot(zeros(Int32, 0), Diagonal(zeros(Int, 0)), zeros(Int16, 0)) === 0
 end
 
+# issue 37359
+@testset "Issue #37359 return Diagonal Matrix for Abstract type :> Number" begin
+    D = Diagonal(rand(Int, n))
+    D2 = Diagonal{Integer}(rand(Int, n))
+    @test isa(I \ Diagonal(ones(Integer, n)), Diagonal)
+    @test D \ D2 ≈ Diagonal(Matrix(D) \ Matrix(D2))
```
Comment (martinholters) [inline stdlib/LinearAlgebra/test/diagonal.jl:734]: Doesn't this test another method? I'd imagine something like
```julia
v = Integer[-5:5...]
@test 2I \ Diagonal(v) == 2\Diagonal(v) == Diagonal(v/2)
```

Comment (tantei3) [reply stdlib/LinearAlgebra/test/diagonal.jl:734]: Added the test you've mentioned above, let me know if you want me to remove the one I added for computation invoking the matrices

Comment (martinholters) [reply stdlib/LinearAlgebra/test/diagonal.jl:734]: :man_shrugging: I thinks it's basically a duplicate of https://github.com/JuliaLang/julia/blob/60c36f39cca453b6875ce8f530cfe0d5657daeae/stdlib/LinearAlgebra/test/diagonal.jl#L192
OTOH, it should not hurt to test this once more, as the time needed is negligible.

Comment (tantei3) [reply stdlib/LinearAlgebra/test/diagonal.jl:734]: Yes but I think the types of D and D2 were being derived from the below line which didn't include Abstract Types relevant for this issue?
```
@testset for relty in (Float32, Float64, BigFloat), elty in (relty, Complex{relty})
```
Please let me know if any other changes are required, happy to learn and improve

**Comment by dkarrasch** [comment]
Superseded by #39701.

**Pull request closed without merging**